csrc/multidevice/utils.cpp

```diff
@@ -142,19 +142,49 @@ int64_t numDeviceDims(TensorView* tv) {
 }
 
 bool isResharding(Expr* expr) {
-  std::unordered_set<TensorView*> tvs;
-  for (auto tv : ir_utils::filterByType<TensorView>(expr->inputs())) {
-    tvs.insert(tv);
-  }
-  for (auto tv : ir_utils::filterByType<TensorView>(expr->outputs())) {
-    tvs.insert(tv);
-  }
-  if (tvs.empty()) {
-    return false;
+  // we don't use getTvsWithDifferentSharding because it creates a computeAtMap,
+  // which is too costly
+  for (auto producer : ir_utils::filterByType<TensorView>(expr->inputs())) {
+    for (auto consumer : ir_utils::filterByType<TensorView>(expr->outputs())) {
+      // exit early in the unsharded case for performance
+      if (!producer->hasDeviceMesh() && !consumer->hasDeviceMesh()) {
+        continue;
+      }
+      // If device mesh are different, the Expr is resharding
+      if (producer->getDeviceMesh().vector() !=
+          consumer->getDeviceMesh().vector()) {
+        return true;
+      }
+
+      // Create a map between producer's and consumer's IterDomains. We iterate
+      // over producer's iterdomain and compare sharding type with consumer's
+      // iterdomain
+      const auto p2c_map =
+          PairwiseRootDomainMap(producer, consumer).mapProducerToConsumer();
+      for (auto p_id :
+           TensorDomain::noReductions(producer->getMaybeRFactorDomain())) {
+        auto p2c_map_it = p2c_map.find(p_id);
+        if (p2c_map_it == p2c_map.end()) {
+          NVF_ERROR(
+              !p_id->isDeviceDim(),
```

Comment:
Isn't this always an error regardless if p_id is device parallelized?

Reply:
I don't know in general but for the case we test you are right! So I made the change accordingly by asserting that all the IDs are mapped to the consumer